torch/distributed/_composable/fsdp/_fsdp_param.py

```diff
@@ -0,0 +1,551 @@
+            assert global_tp_mesh_dim is not None  # mypy
+            # TODO: Hard code inter-node FSDP + intra-node TP; need to support
+            # HSDP and DDP in place of FSDP
+            global_placements[global_dp_mesh_dim] = Shard(0)
+            global_placements[global_tp_mesh_dim] = self._tp_spec.placements[0]
+            self._global_placements = cast(
+                Tuple[Placement, ...], tuple(global_placements)
+            )
+            self._global_size = self._tp_global_size
+            self._global_stride = self._tp_global_stride
+        else:
+            if _mesh_resources.get_parent_mesh(self.mesh_info.mesh) is not None:
+                raise NotImplementedError(
+                    "For pure FSDP/HSDP, using a parent mesh is not supported"
+                )
+            self._global_mesh = self.mesh_info.mesh
+            self._global_placements = (Shard(0),)
+            self._global_size = param.size()
+            self._global_stride = param.stride()
+
+    @torch.no_grad()
+    def _init_sharded_param(self, param: nn.Parameter, device: torch.device):
+        if param.device != device:
+            raise AssertionError(
+                "Expects parameter to already be moved to device "
+                f"{device} but got {param.device}"
+            )
+        param_data = param._local_tensor if isinstance(param, DTensor) else param
+        self._unsharded_size = param_data.size()
+        meta_param = torch.empty(
+            self._unsharded_size,
+            dtype=param.dtype,
+            layout=param.layout,
+            device="meta",
+            requires_grad=param.requires_grad,
+            # Assume contiguous memory layout
+        )
+        shard_rank = self.mesh_info.shard_mesh_rank
+        shard_world_size = self.mesh_info.shard_mesh_size
+        padded_param = pad_tensor_on_dim0(param_data, shard_world_size)
+        padded_chunks = torch.chunk(padded_param, shard_world_size, dim=0)
+        if self.offload_to_cpu:  # ignore `device`
+            self._sharded_param_data = param_data.new_empty(
+                padded_chunks[0].size(), device="cpu"
+            ).pin_memory()
+        else:
+            self._sharded_param_data = param_data.new_empty(
+                padded_chunks[0].size(), device=device
+            )
+        if device != torch.device("meta"):
+            self._sharded_param_data.copy_(padded_chunks[shard_rank])
+        chunks = chunk_with_empty(meta_param, shard_world_size, dim=0)
+        self._sharded_size = chunks[shard_rank].size()
+        self._padded_sharded_size = padded_chunks[0].size()
+        self._padded_unsharded_size = padded_param.size()
+        padded_chunk_numel = padded_chunks[0].numel()
+        self.padded_unsharded_chunk_numels: List[List[int]] = []
+        self.is_padding_mask: List[List[bool]] = []
+        self.unsharded_chunk_numels = []
+        for chunk_idx, chunk in enumerate(chunks):
```

Comment:
This (complex) logic is to compute some metadata for the reduce-scatter copy-in, in particular to be able to `torch.split` the reduce-scatter input (when the gradients are not padded).

One option is to undo this complexity and use the naive Python for loop to slice into the reduce-scatter input (instead of `torch.split`) and rely on the custom copy-in kernel later.